include/dlaf/communication/kernels.h

```diff
@@ -0,0 +1,58 @@
+//
+// Distributed Linear Algebra with Future (DLAF)
+//
+// Copyright (c) 2018-2019, ETH Zurich
+// All rights reserved.
+//
+// Please, refer to the LICENSE file in the root directory.
+// SPDX-License-Identifier: BSD-3-Clause
+//
+
+#pragma once
+
+/// @file
+
+#include <hpx/include/parallel_executors.hpp>
+#include <hpx/modules/threading_base.hpp>
+
+#include <mpi.h>
+
+#include "dlaf/common/assert.h"
+#include "dlaf/common/data.h"
+#include "dlaf/common/pipeline.h"
+#include "dlaf/communication/communicator.h"
+#include "dlaf/communication/executor.h"
+#include "dlaf/communication/message.h"
+#include "dlaf/matrix/tile.h"
+
+namespace dlaf {
+namespace comm {
+
+// Non-blocking sender broadcast
+template <class T>
+void bcast_send(matrix::Tile<const T, Device::CPU> const& tile, common::PromiseGuard<Communicator> pcomm,
+                MPI_Request* req) {
+  auto msg = comm::make_message(common::make_data(tile));
+  MPI_Ibcast(const_cast<T*>(msg.data()), msg.count(), msg.mpi_type(), pcomm.ref().rank(), pcomm.ref(),
+             req);
+}
+
+// Non-blocking receiver broadcast
+template <class T>
+matrix::Tile<const T, Device::CPU> bcast_recv(TileElementSize tile_size, int root_rank,
+                                              common::PromiseGuard<Communicator> pcomm,
+                                              MPI_Request* req) {
```

Comment:
During your work you deleted the `bcast_recv` using an externally allocated tile.
Do you see any problem adding it back?

I'm asking because I think that this solution with the embedded allocation is going to be dropped soon (e.g. with the `Panel` where the matrix will be re-used and not allocated at each communication).

Reply:
@teonnik in the last meeting we had, I was talking about this. Do you have any insight to share about it?